### ISSUE TITLE
Pipes could not be found with Ivy enabled

### ISSUE THREAD

**Issue by alippai** [angular/angular#26436]
# This is an *Ivy* bug report.

## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
Event with the correct imports and declarations I get the following error
```
The pipe 'async' could not be found ("
    </div>

    <div class="progress-bar" [ERROR ->]*ngIf="isProgressBarVisible$ | async"></div>
</div>

"):
```

## Expected behavior
It should compile

## What is the motivation / use case for changing the behavior?
Just checking the compiler status

## Environment

<pre><code>
Angular version: 7.0.0-rc.1
 
For Tooling issues:
- Node version: NodeJS 10.12
- TypeScript version: 3.1.3

Others:
This is an *Ivy* bug report. I reported it, because the status page says pipes should work. Feel free to close this issue, if it's not the right time to submit bug tickets. 
</code></pre>

**Comment by JoostK** [comment]
Hi @alippai,

did you run against 7 RC bundles? Those have not been compiled with Ivy definitions generation, so the pipes won't be available as such. You'd first have to use ngcc to convert pre-Ivy packages to include Ivy definitions, which will also convert `@angular/core`.

You should expect this issue being closed, as no Ivy issues are being taken in at the moment.

**Comment by alippai** [comment]
It emits this error for the pipes defined in the project as well (after removing the async pipe mentioned above).

I used the versions available on npm calling the main() function of the compiler and `enableIvy: true`.

I'll play a little bit more with this, the new build architecture is exciting.

**Comment by JoostK** [comment]
Try `enableIvy: 'ngtsc'` to use the new compiler pipeline.

**Comment by alippai** [comment]
Thanks, I'll try it

**Comment by alippai** [comment]
It's failing for other reasons (first `could not resolve foreign function declaration: app.module.ts forRoot`, after fix `Module not found: [object Object]`).

Debug on Windows fails: `HostResourceLoader: host(C:\Repositories\admin/C:/Repositories/admin/src/ts/components/dashboard/dashboard.component.html) returned a Promise`

So I'll leave this here, maybe it helps the Angular developers.

**Comment by IgorMinar** [comment]
you need to run run ngcc before you compile your code with ivy. checkout some of the ngcc integration tests in the repo. I hope that helps.

in general, there are still too many failing tests for us to resolve before we are ready to accept ivy related issues, so I'm going to close this one.

**Comment by coonmoo** [comment]
Having the same problem, I build with "ivy-ngcc" before running "ng run app:build:production":

`ERROR in could not resolve foreign function declaration: C:/MyApp/src/app/components/components.ts forRoot`

**Comment by alippai** [comment]
I didn't fix it, turned off ivy. It is still developed actively, eg. https://github.com/angular/angular/pull/27737

**Comment by Kajmeh** [comment]
@IgorMinar How can I run ngcc ? I am getting the same error ERROR in could not resolve foreign function declaration app.module.ts forRoot

**Comment by tiagoboeing** [comment]
I have same problem, and turned off ivy too. The problem is with pipes `ERROR in : Template parse errors:
The pipe 'date' could not be found...`.

For turn off `ivy`, open `tsconfig.json` and edit the key ivy to false.

```  
"angularCompilerOptions": {
    "enableIvy": false
}
```

**Comment by leonetosoft** [comment]
Same problem, solved with orientation of @tiagoboeing

**Comment by newmanicspree** [comment]
I'm having the same issue with various pipes being removed with Ivy compiler. Ivy renderer remains `false` in `tsconfig.json` as temporary work-around.

**Comment by dustintheweb** [comment]
Build fail for me also with Ivy enabled 😔
`The pipe 'json' could not be found (" ... "): ...`
Someday i'll get to see this baby in action...

**Comment by frankadrian** [comment]
Turning Ivy off works, but I am interested to know why this works? @tiagoboeing

**Comment by tiagoboeing** [comment]
Ivy is the 'new' compiler for Angular.
I'm not sure, but it may be that the compiler (Ivy) version in your project not recognized @Pipes yet, but on current status Ivy accept the same, second website: https://is-angular-ivy-ready.firebaseapp.com/#/status

Read more: https://blog.angularindepth.com/inside-ivy-exploring-the-new-angular-compiler-ebf85141cee1

**Comment by KSANTHOSHKUMA** [comment]
i am getting the same issue is there any solution to resolve this production build issue.
Thanks in advance.

**Comment by rubenheymans** [comment]
i'm having the same issues using latest angular 8 and angularfirebase

**Comment by RyannGalea** [comment]
@JoostK, Thank you so much. 

I have been pulling my hair out with this issue for months, having to restart my server on every template change was driving me insane. Chasing module imports thinking i have made some sort of error which I hadn't of course.

You have made my Sunday. I moved to 9.0.0-next.5 and it solved it.

**Comment by hypery2k** [comment]
> @ottodranik please test Ivy using the 9.0 preview releases, currently `9.0.0-next.6` is the latest. The issue you describe should have been resolved a while ago, but it's not included in 8.x.

will it be backported to 8.x.x?

**Comment by tuurbo** [comment]
The error ```Error: Uncaught (in promise): Error: The pipe 'async' could not be found!``` is still an issue for me as of 9.0.0-next.7

Happens after I save a module (even without making changes to the module). Restarting ng serve fixes it.

**Comment by JoostK** [comment]
> > @ottodranik please test Ivy using the 9.0 preview releases, currently `9.0.0-next.6` is the latest. The issue you describe should have been resolved a while ago, but it's not included in 8.x.
> 
> will it be backported to 8.x.x?

No, Ivy fixes are only committed to master (i.e. `@next`) while it's in preview. There's no production value in backporting to 8.x.

**Comment by JoostK** [comment]
> The error `Error: Uncaught (in promise): Error: The pipe 'async' could not be found!` is still an issue for me as of 9.0.0-next.7
> 
> Happens after I save a module (even without making changes to the module). Restarting ng serve fixes it.

If this is indeed the case, we'd need to investigate as to why it would still occur. Could you please share the exact versions you're using (e.g. `ng --version`) as that might shed some light on your environment. If nothing seems off in that regard, we'd need a reproduction.

**Comment by tuurbo** [comment]
@JoostK 

```
Angular CLI: 9.0.0-next.5
Node: 12.8.0
OS: win32 x64
Angular: 9.0.0-next.7
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router, service-worker

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-next.5
@angular-devkit/build-angular     0.900.0-next.5
@angular-devkit/build-optimizer   0.900.0-next.5
@angular-devkit/build-webpack     0.900.0-next.5
@angular-devkit/core              9.0.0-next.5
@angular-devkit/schematics        9.0.0-next.5
@angular/cdk                      8.2.0
@angular/cli                      9.0.0-next.5
@angular/material                 8.2.0
@angular/pwa                      0.900.0-next.5
@ngtools/webpack                  9.0.0-next.5
@nguniversal/express-engine       8.1.1
@schematics/angular               9.0.0-next.5
@schematics/update                0.900.0-next.5
rxjs                              6.5.3
typescript                        3.5.3
webpack                           4.40.2
```

**Comment by JoostK** [comment]
@tuurbo thanks, that all seems good to me. Could you please open a new issue and provide a reproduction, preferably by creating a repo in Github?

**Comment by tuurbo** [comment]
@JoostK 

Spent 3 hours trying to recreate the issue.. the best I can do is give you some things I learned.

- The error doesn't seem specific to the async pipe. If I remove async from the template, I get ```Can't bind to 'ngIf' since it isn't a known property of 'ng-container'.``` instead.
- Only the lazy route of the problematic module/component are affected by the errors. I can click on other lazy routes and they load fine.
- Was able to resolve the issue by narrowing down to the problem component, but what's weird is even if i striped it down to a empty module/component that just has a simple string as the component template, the error still persists.
- By deleting the problem module and moving its imports directly into the parent module i was able to fix the error

**Comment by prudnikov** [comment]
This is happening in my project after I edit template where async pipe is used. After restarting the dev server everything back to normal.

Ivy enabled, Angular 8.2.9.

**Comment by klemenoslaj** [comment]
I am facing the  exact same issues...

```
Angular CLI: 9.0.0-next.12
Node: 10.16.0
OS: darwin x64
Angular: 9.0.0-next.11
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.0-next.12
@angular-devkit/build-angular      0.900.0-next.12
@angular-devkit/build-ng-packagr   0.900.0-next.12
@angular-devkit/build-optimizer    0.900.0-next.12
@angular-devkit/build-webpack      0.900.0-next.12
@angular-devkit/core               9.0.0-next.12
@angular-devkit/schematics         9.0.0-next.12
@angular/cdk                       9.0.0-next.0
@angular/cli                       9.0.0-next.12
@angular/flex-layout               8.0.0-beta.27
@angular/localize                  9.0.0-next.5
@angular/material                  9.0.0-next.0
@ngtools/webpack                   9.0.0-next.12
@schematics/angular                9.0.0-next.12
@schematics/update                 0.900.0-next.12
ng-packagr                         5.7.0
rxjs                               6.5.3
typescript                         3.5.3
webpack                            4.41.1
```

**Comment by dmattsson** [comment]
I'm also facing this
``` @angular-devkit/architect         0.900.0-rc.1
@angular-devkit/build-angular     0.900.0-rc.1
@angular-devkit/build-optimizer   0.900.0-rc.1
@angular-devkit/build-webpack     0.900.0-rc.1
@angular-devkit/core              9.0.0-rc.1
@angular-devkit/schematics        9.0.0-rc.1
@angular/cdk                      8.2.3
@angular/fire                     5.2.1
@angular/material                 8.2.3
@ngtools/webpack                  9.0.0-rc.1
@schematics/angular               9.0.0-rc.1
@schematics/update                0.900.0-rc.1
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
```

**Comment by Splaktar** [comment]
@dmattsson can you try with material/cdk `9.0.0-rc.1`? Also does removing `node_modules` and reinstalling, rebuilding solve this?

**Comment by tuurbo** [comment]
@Splaktar 

I made a app to reproduce the error https://github.com/tuurbo/angular-async-error

**Comment by Splaktar** [comment]
@tuurbo if you add `CommonModule` to the `AppModule` does it work? Please open a new issue and include your repo link.

**Comment by tuurbo** [comment]
@Splaktar I added CommonModule as you asked and it didnt fix the issue. I also created a new issue https://github.com/angular/angular/issues/33950

**Comment by tuurbo** [comment]
Well... seems fixed as of angular and cli 9.0.0-rc.3. I just closed the new issue. I'll reopen if i experience it again...

**Comment by neil-119** [comment]
I'm using 9.0.0-rc.3 and I declared a pipe in `AppModule` (the pipe itself resides in our `src/shared/pipes` folder) and it keeps saying could not be found. Now I can't even get the application to run, running into issue #33821.
